src/net/neighbors/walk.rs

```diff
@@ -0,0 +1,2003 @@
+
+        Ok(w)
+    }
+
+    /// Instantiate the neighbor walk to an always-allowed node.
+    /// If we're in the initial block download, then this must also be a *bootstrap* peer.
+    /// Returns the neighbor walk on success
+    /// Returns NotFoundError if no always-allwed neighbors are in the DB.
+    /// Returns DBError if there's a problem querying the DB
+    pub(crate) fn instantiate_walk_to_always_allowed(
+        db: DB,
+        comms: NC,
+        network: &PeerNetwork,
+        ibd: bool,
+    ) -> Result<NeighborWalk<DB, NC>, net_error> {
+        let mut allowed_peers = db.get_initial_walk_neighbors(network, ibd)?;
+        let allowed_peer = if let Some(peer) = allowed_peers.pop() {
+            peer
+        } else {
+            // no allowed peers in DB. Try a different strategy
+            return Err(net_error::NotFoundError);
+        };
+
+        debug!(
+            "Will (re-)connect to always-allowed peer {:?}",
+            &allowed_peer.addr
+        );
+        let w = NeighborWalk::new(
+            db,
+            comms,
+            &allowed_peer,
+            true,
+            network.get_walk_pingbacks().clone(),
+            &network.get_connection_opts(),
+        );
+
+        debug!(
+            "{:?}: instantiated neighbor walk to always-allowed peer {:?}",
+            network.get_local_peer(),
+            &allowed_peer
+        );
+        Ok(w)
+    }
+
+    /// Instantiate a neighbor walk, but use an inbound neighbor instead of a neighbor from our
+    /// peer DB.  This helps a public node discover other public nodes, by asking a private node
+    /// for its neighbors (which can include other public nodes).
+    /// If an inbound connection is found, then return the walk to it.
+    /// Otherwise, return NoSuchNeighbor
+    pub(crate) fn instantiate_walk_from_inbound(
+        db: DB,
+        comms: NC,
+        network: &PeerNetwork,
+    ) -> Result<NeighborWalk<DB, NC>, net_error> {
+        if network.get_num_p2p_convos() == 0 {
+            debug!(
+                "{:?}: failed to begin inbound neighbor walk: no one's connected to us",
+                network.get_local_peer()
+            );
+            return Err(net_error::NoSuchNeighbor);
+        }
+
+        // pick a random search index
+        let mut idx = thread_rng().gen::<usize>() % network.get_num_p2p_convos();
+
+        test_debug!(
+            "{:?}: try inbound neighbors -- sample out of {}. idx = {}",
+            network.get_local_peer(),
+            network.get_num_p2p_convos(),
+            idx
+        );
+
+        // find an inbound connection
+        for _ in 0..network.get_walk_pingbacks().len() + 1 {
+            let event_id = match network.iter_peer_event_ids().skip(idx).next() {
+                Some(eid) => *eid,
+                None => {
+                    idx = 0;
+                    continue;
+                }
+            };
+            idx = (idx + 1) % network.get_num_p2p_convos();
```

Comment:
This loop is confusing to me -- why are you iterating `get_walk_pingbacks().len()` times, but over the number of conversations? Are those items the same length?